varlink/internal/ctxio/conn.go

```diff
@@ -0,0 +1,109 @@
+package ctxio
+
+import (
+	"bufio"
+	"context"
+	"net"
+	"time"
+)
+
+// Conn wraps net.Conn with context aware functionality.
+type Conn struct {
+	conn   net.Conn
+	reader *bufio.Reader
+}
+
+// NewConn creates a new context aware Conn.
+func NewConn(c net.Conn) *Conn {
+	return &Conn{
+		conn:   c,
+		reader: bufio.NewReader(c),
+	}
+}
+
+type wret struct {
+	n   int
+	err error
+}
+
+type rret struct {
+	val []byte
+	err error
+}
+
+// aLongTimeAgo is a time in the past that indicates a connection should
+// immediately time out.
+var aLongTimeAgo = time.Unix(1, 0)
+
+// Close releases the Conns resources.
+func (c *Conn) Close() error {
+	return c.conn.Close()
+}
+
+// Write writes to the underlying connection.
+// It is not safe for concurrent use with itself.
+func (c *Conn) Write(ctx context.Context, buf []byte) (int, error) {
+	// Enable immediate connection cancelation via context by using the context's
+	// deadline and also setting a deadline in the past if/when the context is
+	// canceled. This pattern courtesy of @acln from #networking on Gophers Slack.
+	dl, _ := ctx.Deadline()
+	if err := c.conn.SetWriteDeadline(dl); err != nil {
+		return 0, err
+	}
+
+	ch := make(chan wret, 1)
+	go func() {
+		n, err := c.conn.Write(buf)
```

Comment:
This isn't cancelled if ctx is. So we're blocked...
Also, you should make it clear that we don't leak go routines by the time the function is done with a wg or a comment if you're relying on that ch read. It's particularly troublesome with the buffered channel you're using.

Reply:
`SetWriteDeadline` cancels any current reads. The channel already guarantees that we don't leak goroutines because it's read from before we return.